fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/ratefactor/RateFactorFunctions.java

```diff
@@ -0,0 +1,134 @@
+     *
+     * fn = 1 + fnValueFrom * rateFactorEnd
+     *
+     * @param periodFrom
+     *
+     * @param periodEnd
+     *
+     * @param mathContext
+     *
+     */
+    public static RateFactorPeriodData fnValue(RateFactorPeriodData periodFrom, final RateFactorPeriodData periodEnd,
+            final MathContext mc) {
+        BigDecimal fnValue = BigDecimal.ONE;
+        if (periodEnd.getPeriod() > 1) {
+            fnValue = BigDecimal.ONE.add(periodFrom.getFnValue().multiply(periodEnd.getRateFactor(), mc));
+        }
+        periodEnd.setFnValue(fnValue);
+        return periodEnd;
+    }
+
+    /**
+     *
+     *
+     * emi = (R^n * principal) / fnValue
+     *
+     * @param periods
+     *
+     * @param principal
+     *
+     * @param mathContext
+     *
+     */
+    public static BigDecimal calculateEMI(final List<RateFactorPeriodData> periods, final BigDecimal principal, final MathContext mc,
+            final Integer precision) {
+        final BigDecimal lastFnValue = periods.get(periods.size() - 1).getFnValue();
+        BigDecimal rnValue = BigDecimal.ONE;
+        for (RateFactorPeriodData rateFactorPeriod : periods) {
+            if (rateFactorPeriod.getPeriod() == 1) {
+                rnValue = rateFactorPeriod.getRateFactor();
+            } else {
+                rnValue = rnValue.multiply(rateFactorPeriod.getRateFactor(), mc);
+            }
+        }
+        BigDecimal emiAmount = rnValue.multiply(principal.divide(lastFnValue, mc), mc);
+        return emiAmount.setScale(precision, MoneyHelper.getRoundingMode());
+    }
+
+    public static BigDecimal calculateEMI(final List<LoanRepaymentScheduleInstallment> installments, final BigDecimal principal,
+            final BigDecimal interestRate, final DaysInYearType daysInYearType, final MathContext mc) {
+        final List<RateFactorPeriodData> rateFactorDatas = new ArrayList<>();
+        RateFactorPeriodData rateFactorPeriodDataFrom = null;
+        // Calculate the Rate Factor and fnValue for each Installment
+        for (LoanRepaymentScheduleInstallment installment : installments) {
+            final Long daysInPeriod = DateUtils.getDifferenceInDays(installment.getFromDate(), installment.getDueDate());
+            final Integer daysInYear = RateFactorFunctions.daysInYear(daysInYearType, installment.getFromDate());
+            BigDecimal rateFactor = RateFactorFunctions.rateFactor(interestRate, daysInPeriod, daysInYear, PRECISION);
+
+            RateFactorPeriodData rateFactorPeriodDataEnd = new RateFactorPeriodData(installment.getInstallmentNumber(),
+                    installment.getFromDate(), installment.getDueDate(), rateFactor);
+            RateFactorFunctions.fnValue(rateFactorPeriodDataFrom, rateFactorPeriodDataEnd, mc);
+
+            rateFactorDatas.add(rateFactorPeriodDataEnd);
+            rateFactorPeriodDataFrom = rateFactorPeriodDataEnd;
+        }
+
+        return calculateEMI(rateFactorDatas, principal, mc, 2);
```

Comment:
I am assuming this one is mimicking when we convert it to Money object, but it would be better instead of mimicking actually convert it to Money

Reply:
Done! code removed from this PR